File: cloud/src/core/model_manifest.ts
/**
 * Server-learned model hashes + the platform-signed fetch manifest. Ported
 * from `server/comfyfed_server/model_manifest.py`, read in full -- see that
 * module's docstring for the full consensus/conflict/signing rationale this
 * mirrors.
 *
 * Two pieces:
 *
 * 1. `recordHash` learns a (name, size_bytes) -> sha256 consensus from
 *    `inventory` reports (`do/hub.ts`'s `handleInventory`, mirroring
 *    `agentws._handle_inventory` -> `model_manifest.record_hash`). Two
 *    workers reporting DIFFERENT hashes for the same (name, size_bytes) is a
 *    same-name-different-content collision -- never silently overwritten:
 *    the first-seen hash wins and the row is marked `conflict = true`
 *    (`queries.markModelHashConflict`), excluding it from the manifest.
 *
 * 2. `entries` builds the signed fetch manifest: joins `model_guide`'s
 *    name -> download-source lookup with the learned, non-conflicted hashes
 *    above (`queries.getAllModelHashes` already filters `conflict = 0` in
 *    SQL). Only a model with BOTH a known source URL and an agreed sha256
 *    becomes a manifest entry, each carrying a platform Ed25519 signature
 *    over its own `name|directory|sha256|size_bytes` (see `lib/signing.ts`'s
 *    `buildManifestEntryPayload`).
 *
 * Fix round 1 (Task 7 review, m1): the conflict flag is a persisted D1
 * column (migration 0005_model_hash_conflict.sql), not an in-memory,
 * per-DO-instance set -- the earlier design here couldn't be seen by a
 * plain HTTP route (`routes/workers.ts`'s manifest endpoints, `routes/
 * comfyapi.ts`/`routes/jobs.ts`'s submission gates) with no access to the
 * Hub DO's memory, and was forgotten on DO eviction. Persisting it makes
 * exclusion a plain SQL predicate every caller gets for free, with no
 * coordination and no staleness window.
 *
 * Phase 3.2 addendum, ported from `model_manifest.py`: `entries()` no
 * longer requires a learned consensus row at all for the 11 curated models
 * in `model_guide.SOURCES` that carry an operator-vouched `sha256`/
 * `sizeBytes` -- a "zero-holder" curated model (no worker in the fleet has
 * ever reported it, so no `model_hashes` row can exist) still gets a signed
 * manifest entry, built straight from those guide values (see
 * `guideHashEntry`). A learned consensus, once any worker reports one,
 * always wins over the guide value for that entry; a guide/consensus
 * mismatch is logged (deduped per (name, guide, consensus) triple via
 * `MISMATCH_LOGGED`, a module-level Set -- fine for a single worker isolate,
 * same as the Python process-level set) and the consensus hash is what
 * ships. A conflicted `model_hashes` row still excludes the name outright,
 * guide hash or not -- see `entries()`.
 */

import { matchesModelName } from "./assess";
import * as modelGuide from "./model_guide";
import * as peer from "./peer";
import * as queries from "../db/queries";
import { toSqliteTimestamp } from "../db/queries";
import { signHex } from "../lib/ed25519";
import { buildManifestEntryPayload } from "../lib/signing";

export interface ManifestEntry {
  name: string;
  directory: string;
  /** null only for a Phase 3.1 peer-only entry (no known download source at
   * all, an online seeder instead). */
  url: string | null;
  backup_url: string | null;
  sha256: string;
  size_bytes: number;
  sig: string;
  /** Phase 3.1: present (true) when this (name, size_bytes) also has an
   * online P2P seeder right now -- informative for a URL-sourced entry (the
   * agent fetcher always tries peer first regardless), load-bearing for a
   * peer-only entry (url is null, this is the ONLY source). Absent entirely
   * for a URL-sourced entry with no online seeder, matching peer.py's
   * `entries()` shape (`peer` key only set when true, never `false`). */
  peer?: true;
}

// Guide-vs-consensus mismatch warnings already emitted (dedup; see
// `entries()`) -- module-level Set, ported from Python's `_MISMATCH_LOGGED`.
// Fine to keep per-isolate: a repeat mismatch after an isolate recycle just
// logs once more, no correctness dependency on this surviving a cold start.
const MISMATCH_LOGGED = new Set<string>();

/** Test-only escape hatch (Phase 3.2 F6 fix), ports model_manifest.py's
 * `_clear_mismatch_log_for_tests`: reset the module-level guide-vs-consensus
 * mismatch dedup set so a test asserting a warning was logged isn't broken
 * by test ORDER -- only the first execution of a given (name, guide,
 * consensus) triple in this module instance actually logs otherwise. NOT
 * for production use. */
export function clearMismatchLogForTests(): void {
  MISMATCH_LOGGED.clear();
}

/** The curated guide sha256 for inventory-relative `name`, when it matches
 * one of the 11 `model_guide.SOURCES` entries that carries a vouched hash --
 * else null. Used only to annotate a `recordHash` conflict warning (Phase
 * 3.2 F4 fix) with a diagnostic clue; never to resolve/prefer a value -- a
 * `model_hashes` conflict stays purely reporter-vs-reporter, and the guide
 * never writes to this table. Ports model_manifest.py's `_guide_sha256_for`. */
function guideSha256For(name: string): string | null {
  for (const source of Object.values(modelGuide.SOURCES)) {
    if (source.sha256 && matchesModelName(name, source.name)) return source.sha256;
  }
  return null;
}

export interface RecordHashResult {
  /** True when this report conflicted with an already-learned hash for the
   * same (name, size_bytes) -- the row has already been marked
   * `conflict = true` in D1 by the time this returns; the caller does not
   * need to track anything else. */
  conflict: boolean;
}

/** Learn one inventory entry's sha256 for (name, size_bytes) -- ports
 * `model_manifest.record_hash`. `sizeBytes` must be the EXACT byte count
 * (see that Python docstring for why); the caller (`do/hub.ts`'s
 * `handleInventory`) is responsible for the same GB-rounding fallback
 * `agentws._record_model_hashes` documents.
 *
 * `chunkSha256s` (Phase 3.1 addendum): stored (as JSON text) the FIRST time
 * a reporter's whole-file hash matches/establishes consensus for (name,
 * size_bytes) and no chunk list is on the row yet -- never overwritten by a
 * later, different one (the whole-file hash is the sole trust root; final
 * whole-file verification always runs on every P2P download regardless of
 * chunk hashes). A report that disagrees on the whole-file hash never
 * contributes its chunk list either -- see the conflict branch below. */
export async function recordHash(
  db: D1Database,
  workerId: string,
  name: string,
  sizeBytes: number,
  sha256: string,
  chunkSha256s?: string[] | null
): Promise<RecordHashResult> {
  const chunkJson = chunkSha256s && chunkSha256s.length > 0 ? JSON.stringify(chunkSha256s) : null;

  const inserted = await queries.insertModelHashIfAbsent(
    db,
    name,
    sizeBytes,
    sha256,
    workerId,
    toSqliteTimestamp(new Date()),
    chunkJson
  );
  if (inserted) return { conflict: false };

  const existing = await queries.getModelHash(db, name, sizeBytes);
  if (existing === null) return { conflict: false };

  if (existing.sha256 === sha256) {
    if (chunkJson !== null && existing.chunkSha256s === null) {
      await queries.setModelHashChunksIfAbsent(db, name, sizeBytes, chunkJson);
    } else if (chunkJson !== null && JSON.stringify(existing.chunkSha256s) !== chunkJson) {
      console.warn(
        `model_manifest: chunk_sha256s mismatch for ${name} (size_bytes=${sizeBytes}) reported by worker ` +
          `${workerId} -- whole-file sha256 still agrees, so this is not a conflict; keeping the first-seen ` +
          `chunk list (final whole-file verification is the trust root regardless of any chunk list)`
      );
    }
    return { conflict: false };
  }

  // Phase 3.2 F4 fix: when either side of the conflict happens to equal the
  // curated guide's vouched hash for this name, say so -- that's exactly
  // the clue an operator needs to diagnose "this is a stale registry entry,
  // not two genuinely different worker builds" at a glance.
  const guideSha256 = guideSha256For(name);
  let guideClue = "";
  if (guideSha256 !== null) {
    if (guideSha256 === sha256) {
      guideClue = ` (NOTE: worker ${workerId}'s reported hash matches the curated guide value -- the OTHER report looks stale)`;
    } else if (guideSha256 === existing.sha256) {
      guideClue = ` (NOTE: the existing hash matches the curated guide value -- worker ${workerId}'s NEW report looks stale)`;
    }
  }

  console.warn(
    `model_manifest: sha256 conflict for ${name} (size_bytes=${sizeBytes}): ` +
      `worker ${workerId} reported ${sha256}, worker ${existing.firstWorkerId} previously reported ` +
      `${existing.sha256} -- keeping the first-seen hash and excluding this name from the fetch manifest${guideClue}`
  );
  await queries.markModelHashConflict(db, name, sizeBytes);
  return { conflict: true };
}

/** First `model_hashes` row whose (inventory-relative) name matches
 * `sourceKey` (a bare model_guide name), per `matchesModelName` semantics --
 * ports `model_manifest._find_hash_row`. */
function findHashRow(rows: queries.ModelHashRow[], sourceKey: string): queries.ModelHashRow | null {
  for (const row of rows) {
    if (matchesModelName(row.name, sourceKey)) return row;
  }
  return null;
}

/** Split an inventory-relative model path (`"directory/name"`, the shape a
 * `model_hashes.name` stores) into `(directory, name)` -- ports
 * `model_manifest._split_inventory_name`, the same one-level convention the
 * agent-side peer name sanitizer caps names at. A path with no `/` at all
 * yields `("", inventoryName)`. */
function splitInventoryName(inventoryName: string): [directory: string, name: string] {
  const normalized = inventoryName.replace(/\\/g, "/").replace(/^\/+/, "");
  const slash = normalized.indexOf("/");
  if (slash === -1) return ["", normalized];
  return [normalized.slice(0, slash), normalized.slice(slash + 1)];
}

/** Phase 3.2 zero-holder entry: sign a manifest entry straight from
 * `source`'s operator-vouched `sha256`/`sizeBytes` when NO `model_hashes`
 * row exists for it yet (nobody in the fleet holds the file, so no learned
 * consensus is even possible) -- ports `model_manifest._guide_hash_entry`.
 * Caller (`entries()`) only reaches here after confirming no row --
 * conflicted or otherwise -- matches this name; a learned consensus, once
 * any worker reports one, always takes over from this path (see
 * `entries()`'s consensus-precedence branch).
 *
 * `peer` is deliberately never set on this entry: by construction there is
 * no `model_hashes` row, so there is nothing for `rowHasSeeder` to have
 * matched -- an entry from this path always means zero holders right now. */
async function guideHashEntry(seedHex: string, source: modelGuide.ModelSource): Promise<ManifestEntry | null> {
  if (source.sha256 === undefined || source.sizeBytes === undefined) return null;

  if (source.name.includes("|") || source.directory.includes("|")) {
    console.warn(
      `model_manifest: skipping guide-hash manifest candidate with a '|' in name or directory ` +
        `(payload delimiter): name=${JSON.stringify(source.name)} directory=${JSON.stringify(source.directory)}`
    );
    return null;
  }

  const payload = buildManifestEntryPayload(source.name, source.directory, source.sha256, source.sizeBytes);
  const sig = await signHex(seedHex, new TextEncoder().encode(payload));

  return {
    name: source.name,
    directory: source.directory,
    url: source.officialUrl,
    backup_url: source.backupUrl,
    sha256: source.sha256,
    size_bytes: source.sizeBytes,
    sig,
  };
}

/** Build a peer-only manifest entry for `row` (a non-conflicted
 * `model_hashes` row with no known download source), or null when it has no
 * online seeder right now -- ports `model_manifest._peer_only_entry`. `url`/
 * `backup_url` are null and `peer: true` marks the entry so a consumer
 * (`core/assess.ts`'s protocol>=4 gate, the agent fetcher) knows this model
 * has no URL fallback at all. */
async function peerOnlyEntry(
  seedHex: string,
  row: queries.ModelHashRow,
  seederFiles: ReadonlySet<string>
): Promise<ManifestEntry | null> {
  if (!peer.rowHasSeeder(row, seederFiles)) return null;

  const [directory, name] = splitInventoryName(row.name);

  // Same defensive `|` guard as the URL-sourced branch below -- an agent-
  // reported inventory name is untrusted input relative to this process.
  if (name.includes("|") || directory.includes("|")) {
    console.warn(
      `model_manifest: skipping peer-only manifest candidate with a '|' in name or directory ` +
        `(payload delimiter): name=${JSON.stringify(name)} directory=${JSON.stringify(directory)}`
    );
    return null;
  }

  const payload = buildManifestEntryPayload(name, directory, row.sha256, row.sizeBytes);
  const sig = await signHex(seedHex, new TextEncoder().encode(payload));

  return {
    name,
    directory,
    url: null,
    backup_url: null,
    sha256: row.sha256,
    size_bytes: row.sizeBytes,
    sig,
    peer: true,
  };
}

/** The subset of `entries()`'s output whose ONLY source is a peer (`url` is
 * null) -- names a candidate fetching worker must be protocol>=4 to pull,
 * per the task brief's eligibility gate. A model that has both a URL and an
 * online seeder (`peer: true` alongside a real `url`) is NOT in this set --
 * ports `model_manifest.peer_only_names`. */
export function peerOnlyNames(manifestEntries: ManifestEntry[]): ReadonlySet<string> {
  return new Set(manifestEntries.filter((e) => e.url === null).map((e) => e.name));
}

/** Build the signed fetch-manifest entry list -- ports `model_manifest.
 * entries()`. `queries.getAllModelHashes` already excludes conflicted rows
 * in SQL, so there is no in-memory set for this function (or its caller) to
 * consult.
 *
 * Two kinds of entries (Phase 3.1): URL-sourced (unchanged criteria),
 * additionally carrying `peer: true` when an online seeder also exists right
 * now; and peer-only (`peerOnlyEntry`, above) for a non-conflicted hash row
 * with NO known download source that DOES have an online seeder. */
export async function entries(db: D1Database, store: R2Bucket, seedHex: string): Promise<ManifestEntry[]> {
  const harvested = await modelGuide.harvest(store);
  const names = new Set([...Object.keys(modelGuide.SOURCES), ...Object.keys(harvested)]);

  const hashRows = await queries.getAllModelHashes(db);
  // Phase 3.2: a conflicted row must still block the guide-hash fallback
  // below -- "two reporters disagree" is not resolved by the operator's
  // curated hash, so a name with one stays excluded exactly like before this
  // phase.
  const conflictedRows = await queries.getConflictedModelHashes(db);

  // M3 final-review fix: one worker query + one inventory parse per worker
  // for this whole call, instead of once per hash row below.
  const seederFiles = await peer.seederCandidateFiles(db);

  const result: ManifestEntry[] = [];
  const usedRows = new Set<string>();
  for (const name of [...names].sort()) {
    const source = await modelGuide.lookup(name, store);
    if (source === null || !source.officialUrl) continue;

    const row = findHashRow(hashRows, name);
    if (row === null) {
      // Phase 3.2: no learned consensus for this (curated or harvested)
      // source. A conflicted row still excludes it outright -- consensus
      // precedence never lets a curated guide hash paper over a genuine
      // reporter disagreement.
      if (findHashRow(conflictedRows, name) !== null) continue;
      const guideEntry = await guideHashEntry(seedHex, source);
      if (guideEntry !== null) result.push(guideEntry);
      continue;
    }

    if (source.sha256 !== undefined && source.sha256 !== row.sha256) {
      const dedupKey = `${source.name} ${source.sha256} ${row.sha256}`;
      // Dedup per (name, pair): entries() can be read repeatedly (every
      // manifest fetch), and a standing mismatch would otherwise flood the
      // log.
      if (!MISMATCH_LOGGED.has(dedupKey)) {
        MISMATCH_LOGGED.add(dedupKey);
        console.warn(
          `model_manifest: guide sha256 for ${source.name} differs from the learned consensus ` +
            `(guide=${source.sha256} consensus=${row.sha256}) -- consensus wins, the manifest entry uses the ` +
            `learned hash`
        );
      }
    }

    // Defensive: `|` is the field delimiter in the signed payload below --
    // see model_manifest.py's docstring for why a harvested entry's
    // name/directory (untrusted, off a workflow JSON) must be refused rather
    // than let a crafted value shift what substring the signature covers.
    if (source.name.includes("|") || source.directory.includes("|")) {
      console.warn(
        `model_manifest: skipping manifest candidate with a '|' in name or directory ` +
          `(payload delimiter): name=${JSON.stringify(source.name)} directory=${JSON.stringify(source.directory)}`
      );
      continue;
    }

    const payload = buildManifestEntryPayload(source.name, source.directory, row.sha256, row.sizeBytes);
    const sig = await signHex(seedHex, new TextEncoder().encode(payload));

    const entry: ManifestEntry = {
      name: source.name,
      directory: source.directory,
      url: source.officialUrl,
      backup_url: source.backupUrl,
      sha256: row.sha256,
      size_bytes: row.sizeBytes,
      sig,
    };
    if (peer.rowHasSeeder(row, seederFiles)) {
      entry.peer = true;
    }
    result.push(entry);
    usedRows.add(`${row.name} ${row.sizeBytes}`);
  }

  // Phase 3.1: every remaining non-conflicted hash row (no known download
  // source at all) becomes a peer-only entry when it has an online seeder
  // right now.
  for (const row of hashRows) {
    if (usedRows.has(`${row.name} ${row.sizeBytes}`)) continue;
    const peerEntry = await peerOnlyEntry(seedHex, row, seederFiles);
    if (peerEntry !== null) result.push(peerEntry);
  }

  return result;
}

File: server/comfyfed_server/model_manifest.py
"""Server-learned model hashes + the platform-signed fetch manifest.

Phase 2.1 Task 2 (model auto-distribution). Agents now optionally hash their
local model files and report `sha256` alongside `name`/`size` in their
`inventory` WS message (Phase 2.1 Task 1, `comfyfed_agent.hardware.scan_models`).
This module:

1. Learns a (name, size_bytes) -> sha256 consensus from those reports
   (`record_hash`, called from `agentws._handle_inventory` for every entry
   that carries a `sha256`). Two workers reporting DIFFERENT hashes for the
   same (name, size_bytes) is a same-name-different-content collision --
   never silently overwritten. The first-seen hash wins, a WARNING names
   both workers, and the model name is "poisoned" (excluded from the
   manifest) for the rest of this process's lifetime.

2. Builds the signed fetch manifest (`entries`): joins `model_guide`'s
   name -> download-source lookup (curated `SOURCES` first, `harvest()`
   fallback -- see model_guide.py) with the learned hashes above. Only a
   model with BOTH a known source URL and an agreed sha256 becomes a
   manifest entry, each carrying a platform Ed25519 signature over its own
   `name|directory|sha256|size_bytes` so an agent (or `comfyfed_server`
   dispatch code) can trust a fetched file without re-deriving trust from
   the HTTP connection it came over.

`size_bytes` is the inventory entry's EXACT `os.stat().st_size` (the
`size_bytes` field `hardware.scan_models` adds alongside the display/
compat `size` GB figure -- see its docstring), NOT `model_guide.
ModelSource.size_gb` (hand-curated, coarser, and describes the *expected*
download, not what a worker actually measured on disk) and NOT reconstructed
from the rounded-to-3dp `size` GB figure -- the signed trust payload must
pin the real byte length. An agent that hashes but predates the exact
`size_bytes` field is the one exception: `agentws._record_model_hashes`
falls back to `round(size * 1024**3)` for those reports and documents the
approximation there, at the one call site that needs it, rather than here.

A hash conflict is recorded directly on the `model_hashes` row (`conflict`,
added by migration `c9d0e1f2a3b4`) rather than in an in-memory, per-process
set -- this is a persistent replacement for an earlier in-memory
`_poisoned_names` design, which forgot every conflict on restart and had no
way to be consulted by a second replica behind a load balancer. `entries()`
now excludes any row with `conflict=True` as a plain SQL predicate: no
process/replica coordination needed, and a restart changes nothing about
which names are excluded.
"""

from __future__ import annotations

import json
import logging
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends

from . import assess, auth, db, model_guide, peer, security, workers

logger = logging.getLogger(__name__)


def _utcnow() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def record_hash(
    worker_id: str,
    name: str,
    size_bytes: int,
    sha256: str,
    chunk_sha256s: list[str] | None = None,
) -> None:
    """Learn one inventory entry's sha256 for (name, size_bytes), and (Phase
    3.1 addendum) its per-64-MiB-chunk hash list alongside it.

    `size_bytes` must be the EXACT byte count (an agent's `os.stat().
    st_size`, per `hardware.scan_models`'s `size_bytes` field) -- this is
    what the manifest signs, so a caller reconstructing it from the
    coarser, rounded-to-3-decimal-GB `size` field (only needed for an
    agent that hashes but predates exact `size_bytes` -- see
    `agentws._record_model_hashes`) must do so itself and document the
    approximation at that call site, not here.

    INSERT OR IGNORE semantics: a first report for (name, size_bytes) is
    stored as-is; a later report for the same key with a MATCHING sha256 is
    a no-op; a later report with a DIFFERENT sha256 is a genuine conflict --
    logged at WARNING with both worker ids, the existing row's sha256/
    first_worker_id are left untouched, and its `conflict` column is set to
    True so `entries()` excludes it. Persistent, not in-memory: unlike the
    old per-process poisoned-name set, this survives a restart and is
    visible to every replica immediately (see this module's docstring).

    `chunk_sha256s` (Phase 3.1 addendum, `db.ModelHash.chunk_sha256s`):
    stored (as JSON text) the FIRST time a reporter's whole-file hash
    matches/establishes consensus for (name, size_bytes) and no chunk list
    is on the row yet. Once a chunk list is stored it is never overwritten
    by a later, different one -- the whole-file hash is the sole trust
    root (final whole-file verification always runs on every P2P download
    regardless of chunk hashes), so a second reporter's differing chunk
    list is merely suspicious, not a poisoning event: logged once at
    WARNING and otherwise ignored, WITHOUT touching `conflict` (that stays
    reserved for a genuine whole-file sha256 disagreement). A report that
    disagrees on the whole-file hash never contributes its chunk list
    either -- see the conflict branch below.
    """
    chunk_json = json.dumps(chunk_sha256s) if chunk_sha256s else None

    with db.get_session() as session:
        existing = session.get(db.ModelHash, (name, size_bytes))
        if existing is None:
            session.add(
                db.ModelHash(
                    name=name,
                    size_bytes=size_bytes,
                    sha256=sha256,
                    first_worker_id=worker_id,
                    created_at=_utcnow(),
                    chunk_sha256s=chunk_json,
                )
            )
            session.commit()
            return

        if existing.sha256 == sha256:
            if chunk_json is not None:
                if existing.chunk_sha256s is None:
                    existing.chunk_sha256s = chunk_json
                    session.commit()
                elif existing.chunk_sha256s != chunk_json:
                    logger.warning(
                        "model_manifest: chunk_sha256s mismatch for %s "
                        "(size_bytes=%s) reported by worker %s -- whole-file "
                        "sha256 still agrees, so this is not a conflict; "
                        "keeping the first-seen chunk list (final whole-file "
                        "verification is the trust root regardless of any "
                        "chunk list)",
                        name,
                        size_bytes,
                        worker_id,
                    )
            return

        logger.warning(
            "model_manifest: sha256 conflict for %s (size_bytes=%s): "
            "worker %s reported %s, worker %s previously reported %s -- "
            "keeping the first-seen hash and excluding this name from the "
            "fetch manifest",
            name,
            size_bytes,
            worker_id,
            sha256,
            existing.first_worker_id,
            existing.sha256,
        )
        existing.conflict = True
        session.commit()


def _find_hash_row(rows: list, source_key: str):
    """First `model_hashes` row whose (inventory-relative) name matches
    `source_key` (a bare model_guide name), per `assess.matches_model_name`
    semantics -- same name-normalization an inventory entry uses to satisfy
    a workflow's loader value.
    """
    for row in rows:
        if assess.matches_model_name(row.name, source_key):
            return row
    return None


def _split_inventory_name(inventory_name: str) -> tuple[str, str]:
    """Split an inventory-relative model path (`"directory/name"`, the shape
    `hardware.scan_models` reports and `db.ModelHash.name` stores) into
    `(directory, name)`, the same one-level convention Task 5's agent-side
    `fetcher._inventory_name` reconstructs from and `peerserve._sanitize_peer_name`
    caps names at. A path with no `/` at all (no category) yields
    `("", inventory_name)`.
    """
    normalized = inventory_name.replace("\\", "/").lstrip("/")
    directory, separator, remainder = normalized.partition("/")
    if not separator:
        return "", normalized
    return directory, remainder


def _seeder_candidate_files(session) -> frozenset[tuple[str, int, str]]:
    """Every `(name, size_bytes, sha256)` triple currently offered by an
    online, protocol>=4, peer_url-advertising worker -- built ONCE per
    `entries()` call (final-review M3 fix) so the per-`model_hashes`-row loop
    below can answer "does this row have an online seeder right now?" with an
    O(1) set-membership check instead of calling `peer.online_seeders` (its
    own `db.Worker` query plus a fresh `json.loads` of every worker's
    `model_inventory`) once per row. Before this, R hash rows and W workers
    meant R worker queries and R x W inventory parses every `entries()` call
    (every 5s dispatch tick, per-request from jobs.py/comfyapi.py) -- this
    hoists the worker query and every inventory parse out of that loop
    entirely, to run exactly once.

    Same predicate as `peer.online_seeders` (online, protocol>=4, peer_url
    not null), reused as a plain SQL filter here since the module already
    defines `_MIN_PEER_PROTOCOL`, rather than re-deriving it.
    """
    query = (
        session.query(db.Worker)
        .filter(db.Worker.status != "offline")
        .filter(db.Worker.protocol >= peer._MIN_PEER_PROTOCOL)
        .filter(db.Worker.peer_url.isnot(None))
    )
    files: set[tuple[str, int, str]] = set()
    for worker in query.all():
        for inv_entry in peer._worker_inventory(worker):
            if not isinstance(inv_entry, dict):
                continue
            name = inv_entry.get("name")
            size_bytes = inv_entry.get("size_bytes")
            sha256 = inv_entry.get("sha256")
            if not isinstance(name, str) or not isinstance(sha256, str):
                continue
            if not isinstance(size_bytes, int) or isinstance(size_bytes, bool):
                continue
            files.add((name, size_bytes, sha256))
    return files


def _row_has_seeder(row: db.ModelHash, seeder_files: frozenset[tuple[str, int, str]]) -> bool:
    """Whether `row` (name, size_bytes, sha256) is in the batched
    `_seeder_candidate_files()` set -- the per-row replacement for calling
    `peer.online_seeders(session, row.name, row.size_bytes)` (see M3)."""
    return (row.name, row.size_bytes, row.sha256) in seeder_files


def _peer_only_entry(signing_key, row: db.ModelHash, seeder_files: frozenset[tuple[str, int, str]]) -> Optional[dict]:
    """Build a peer-only manifest entry for `row` (a non-conflicted
    `model_hashes` row with no known download source), or None when it has no
    online seeder right now.

    This is `entries()`'s embodiment of the task brief's `fetchable(name,
    size)` predicate for the peer branch: "consensus hash exists AND at least
    one online seeder offers it" -- `seeder_files`, the batched set built
    once per `entries()` call by `_seeder_candidate_files` (M3), replaces a
    direct `peer.online_seeders` call here so this stays a cheap set lookup
    per row instead of its own DB query + inventory parse.
    `url`/`backup_url` are None and `peer: True` marks the entry so a
    consumer (assess.verdict's protocol>=4 gate, the agent fetcher) knows
    this model has no URL fallback at all.
    """
    if not _row_has_seeder(row, seeder_files):
        return None

    directory, name = _split_inventory_name(row.name)

    # Same defensive `|` guard as the URL-sourced branch below -- an agent-
    # reported inventory name is untrusted input relative to this process.
    if "|" in name or "|" in directory:
        logger.warning(
            "model_manifest: skipping peer-only manifest candidate with a "
            "'|' in name or directory (payload delimiter): name=%r directory=%r",
            name,
            directory,
        )
        return None

    payload = f"{name}|{directory}|{row.sha256}|{row.size_bytes}"
    sig = signing_key.sign(payload.encode()).signature.hex()

    return {
        "name": name,
        "directory": directory,
        "url": None,
        "backup_url": None,
        "sha256": row.sha256,
        "size_bytes": row.size_bytes,
        "sig": sig,
        "peer": True,
    }


def peer_only_names(manifest_entries: list[dict]) -> frozenset[str]:
    """The subset of `entries()`'s output whose ONLY source is a peer (`url`
    is None) -- names a candidate fetching worker must be protocol>=4 to
    pull, per the task brief's eligibility gate. A model that has both a URL
    and an online seeder (`peer: True` alongside a real `url`) is NOT in this
    set: protocol>=3 stays sufficient for it, exactly as today, since the
    agent fetcher falls back to the URL chain when a peer pull fails.
    """
    return frozenset(e["name"] for e in manifest_entries if e.get("url") is None)


def entries(data_dir: str) -> list[dict]:
    """Build the signed fetch-manifest entry list.

    Two kinds of entries:

    1. URL-sourced: a known download source (curated or harvested, with a
       non-empty `official_url`) AND an agreed, non-conflicting learned
       sha256 (`model_hashes.conflict == False`). If that same (name,
       size_bytes) also has an online peer seeder right now
       (`peer.online_seeders`), the entry additionally carries `"peer":
       True` -- informative only, since the agent fetcher always tries a
       peer source first regardless of this flag (Task 5) -- so dispatch/
       assess never need a second code path to learn "this model also has a
       seeder".
    2. Peer-only (Phase 3.1): a `model_hashes` row with an agreed hash but NO
       known download source, that DOES have an online seeder right now --
       see `_peer_only_entry`. `url`/`backup_url` are None and `peer: True`.

    Every entry's `sig` is a platform Ed25519 signature (hex) over
    `f"{name}|{directory}|{sha256}|{size_bytes}"` -- unchanged shape for both
    kinds, the URL fields are simply not part of what's signed.
    """
    signing_key, _ = security.load_platform_keys(data_dir)

    names = set(model_guide.SOURCES.keys()) | set(model_guide.harvest(data_dir).keys())

    with db.get_session() as session:
        # Persistent exclusion (migration c9d0e1f2a3b4): a conflicted row is
        # never a candidate, full stop -- no in-memory set to consult, and
        # this is correct across a restart or another replica immediately.
        hash_rows = list(
            session.query(db.ModelHash).filter(db.ModelHash.conflict == False).all()  # noqa: E712
        )

        # M3 final-review fix: one worker query + one inventory parse per
        # worker for this whole call, instead of once per hash row below.
        seeder_files = _seeder_candidate_files(session)

        result: list[dict] = []
        used_rows: set[tuple[str, int]] = set()
        for name in sorted(names):
            source = model_guide.lookup(name, data_dir)
            if source is None or not source.official_url:
                continue

            row = _find_hash_row(hash_rows, name)
            if row is None:
                continue

            # Defensive: `|` is the field delimiter in the signed payload
            # below. `source.name`/`source.directory` should never
            # legitimately contain one (model filenames and category folders
            # don't use it), but a harvested entry's `name`/`directory` come
            # straight off a workflow JSON on disk -- untrusted input
            # relative to this process. Letting one through would let a
            # crafted `directory` value (e.g. containing an extra
            # `|1234|sha|`) shift which substring the signature is later
            # parsed as covering, forging a payload the platform never
            # actually intended to sign for.
            if "|" in source.name or "|" in source.directory:
                logger.warning(
                    "model_manifest: skipping manifest candidate with a '|' in "
                    "name or directory (payload delimiter): name=%r directory=%r",
                    source.name,
                    source.directory,
                )
                continue

            payload = f"{source.name}|{source.directory}|{row.sha256}|{row.size_bytes}"
            sig = signing_key.sign(payload.encode()).signature.hex()

            entry = {
                "name": source.name,
                "directory": source.directory,
                "url": source.official_url,
                "backup_url": source.backup_url,
                "sha256": row.sha256,
                "size_bytes": row.size_bytes,
                "sig": sig,
            }
            if _row_has_seeder(row, seeder_files):
                entry["peer"] = True
            result.append(entry)
            used_rows.add((row.name, row.size_bytes))

        # Phase 3.1: every remaining non-conflicted hash row (no known
        # download source at all) becomes a peer-only entry when it has an
        # online seeder right now.
        for row in hash_rows:
            key = (row.name, row.size_bytes)
            if key in used_rows:
                continue
            peer_entry = _peer_only_entry(signing_key, row, seeder_files)
            if peer_entry is not None:
                result.append(peer_entry)

    return result


def create_router(data_dir: str) -> APIRouter:
    """`/api/agent/manifest` (signed-agent auth) and `/api/models/manifest`
    (admin session auth) -- both return `{"entries": [...]}` from `entries()`.
    The agent route is also called internally by dispatch/auto-fetch code
    (Task 3/4), not just over HTTP.
    """
    r = APIRouter()

    @r.get("/api/agent/manifest")
    def agent_manifest(_worker: db.Worker = Depends(workers.verify_agent)):
        return {"entries": entries(data_dir)}

    @r.get("/api/models/manifest")
    def admin_manifest(_payload: dict = Depends(auth.require_admin)):
        return {"entries": entries(data_dir)}

    return r
